controllers/machine_controller_phases.go

```diff
@@ -172,6 +173,10 @@ func (r *MachineReconciler) reconcileBootstrap(ctx context.Context, cluster *clu
 	}
 	bootstrapConfig := externalResult.Result
 
+	// Report a summary of current status of the bootstrap object defined for this machine.
+	bootstrapReadyCondition := conditions.Mirror(conditions.UnstructuredGetter(bootstrapConfig), clusterv1.BootstrapReadyCondition)
+	conditions.Set(m, bootstrapReadyCondition)
+
```

Comment:
I wouldn't necessarily expect it to happen, but if we plan on eventually relying on conditions rather than the existing status fields for logic down the line, should we attempt to detect any discrepancies between conditions and the status fields and throw an error if they do not align?

Reply:
🤔 I really hope that doesn't happen, seems it would be definitely an implementation issue although not sure if this is the right time/place for conditions to solve that problem, wdyt?

Reply:
I'm just wondering what potential downstream consequences we are looking at if semantics start varying widely between our current contract and conditions-based contracts.

If we pro-actively try to align semantics (maybe through warning instead of erroring to start with), then avoid having to try to reconcile later after implementations are already in place.

Alternatively, we could allow for new providers to implement only conditions rather than both the status fields and conditions, and update the logic accordingly:
- If status fields exist, but not conditions, use those
- If conditions exist, but not status fields, use those
- If both status fields and conditions exist, test for alignment and warn/error

Reply:
For the last bullet, would we have to write special logic in every controller?

Reply:
> For the last bullet, would we have to write special logic in every controller?

I think there are possibly two aspects to it:
- Producing controllers: Ensure tests in place verifying that they don't generate conflicting signals
- Consuming controllers (assuming warning rather than error): Verify if both status fields and conditions present that they do not contradict each other, warn (so that hopefully the bug report makes it up to fix the behavior) in the producing controller, then proceed with logic based on status fields (to not break existing behavior)

We could probably find a way to generalize the consistency check, but at the end of the day matching contractual behaviors between consumed status fields vs consumed conditions will likely be controller specific.

Reply:
These are definitely valid points / concerns, let's chat later all of us and make sure to capture action items in an issue. It's also a matter of documentation for folks implementing conditions, pretty important to call it out explicitly.

Reply:
My major concern here is that conditions are intended to express the operational state of a component in the cluster, while flags were designed for driving the internal state machines of controllers. So I do expect things to naturally drift away.
So IMO it is more a matter of naming so the user does not get tricked by similar names

Reply:
I have opened https://github.com/kubernetes-sigs/cluster-api/pull/3152 to start the discussion about what we want to do about conditions-based contracts vs flags.